client/fetch_tagged_state.go

```diff
@@ -0,0 +1,471 @@
+		// can early terminate
+		f.fatalError = fatalError
+		f.Signal()
+	} else if f.numShardsPending == 0 {
+		// success case
+		f.Signal()
+	} else if f.numHostsPending == 0 && f.numShardsPending != 0 {
+		// ie we've received all responses but still weren't able to satisfy all shards, so we need to fail
+		f.fatalError = fmt.Errorf(
+			"received all responses but were unable to satisfy the consistency requirements on %d shards [ err = %v ]",
+			f.numShardsPending, f.errors.FinalError())
+		f.Signal()
+	}
+}
+
+func (f *fetchTaggedState) asIndexQueryResults() (index.QueryResults, error) {
+	f.Lock()
+	defer f.Unlock()
+	if err := f.fatalError; err != nil {
+		return index.QueryResults{}, err
+	}
+
+	limit := maxInt
+	if l := f.op.request.Limit; l != nil {
+		limit = int(*l)
+	}
+
+	return f.results.asIndexQueryResults(limit), nil
+}
+
+func (f *fetchTaggedState) asEncodingSeriesIterators(
+	slicesIterPool *readerSliceOfSlicesIteratorPool,
+	multiIterPool encoding.MultiReaderIteratorPool,
+	seriesIterPool encoding.SeriesIteratorPool,
+	mutableSeriesItersPool encoding.MutableSeriesIteratorsPool,
+) (encoding.SeriesIterators, bool, error) {
+	f.Lock()
+	defer f.Unlock()
+	if err := f.fatalError; err != nil {
+		return nil, false, err
+	}
+
+	limit := maxInt
+	if l := f.op.request.Limit; l != nil {
+		limit = int(*l)
+	}
+
+	iters, exhaustive := f.results.asEncodingSeriesIterators(limit, slicesIterPool,
+		multiIterPool, seriesIterPool, mutableSeriesItersPool)
+
+	return iters, exhaustive, nil
+}
+
+type fetchTaggedShardConsistencyResult struct {
+	enqueued int8
+	success  int8
+	errors   int8
+	done     bool
+}
+
+type fetchTaggedIDResults []*rpc.FetchTaggedIDResult_
+
+// accumulate responses in a map: id -> []{id, tags, data}
+// NB: the value type is a slice to capture results per shard replica
+type fetchTaggedIDResultByIDMap map[string]fetchTaggedIDResults
+
+type fetchTaggedStatePool struct {
+	pool pool.ObjectPool
+}
+
+func newFetchTaggedStatePool(
+	opts pool.ObjectPoolOptions,
+) *fetchTaggedStatePool {
+	p := pool.NewObjectPool(opts)
+	return &fetchTaggedStatePool{pool: p}
+}
+
+func (p *fetchTaggedStatePool) Init() {
+	p.pool.Init(func() interface{} {
+		f := &fetchTaggedState{}
+		f.reset()
+		return f
+	})
+}
+
+func (p *fetchTaggedStatePool) Get() *fetchTaggedState {
+	return p.pool.Get().(*fetchTaggedState)
+}
+
+func (p *fetchTaggedStatePool) Put(f *fetchTaggedState) {
+	p.pool.Put(f)
+}
+
+// TODO(prateek): delete the nolint below once we finish implementation
+// nolint
+type fetchTaggedResultAccumulator struct {
+	// NB(prateek): a fetchTagged request requires we fan out to each shard in the
+	// topology. As a result, we model the response consistency per shard.
+	// Length of this slice == 1 + max shard id in topology
+	shardResults []fetchTaggedShardConsistencyResult // TODO(prateek): should this be moved to the fetchTaggedState(?)
+
+	responses  fetchTaggedIDResultByIDMap
```

Comment:
talked offline: going to try using `[]*rpc.FetchTaggedIDResult_` + sorting instead of this map. It'll be much easier for pooling, allow stable ordered iteration, not require hashing, etc.

Reply:
done